swaybar/tray/tray.c

```diff
@@ -90,9 +91,17 @@ void destroy_tray(struct swaybar_tray *tray) {
 }
 
 void tray_in(int fd, short mask, void *data) {
-	sd_bus *bus = data;
+	struct swaybar *bar = data;
 	int ret;
-	while ((ret = sd_bus_process(bus, NULL)) > 0) {
+
+	if (mask & (POLLHUP | POLLERR)) {
+        sway_log(SWAY_ERROR, "D-Bus connection closed unexpectedly");
+		bar_teardown(bar);
```

Comment:
`bar_teardown()` is already called when the main loop exits, no?

Reply:
Yes. Just setting it to not running should be enough.